### PR TITLE
Stack download workspace Initial

### PR DESCRIPTION
**Pull request by alinxie**

### CODE REVIEW

File: databricks_cli/stack/api.py [hunk 1]
```diff
@@ -139,6 +162,84 @@ def deploy_config(self, stack_config, stack_status=None, **kwargs):
 
         return new_stack_status
 
+    def download_config(self, stack_config, **kwargs):
```
Comment (k4jiang) [inline databricks_cli/stack/api.py]: Minor: 
1) Nomenclature is a bit off. It sounds as if you're downloading the config. 
2) Consider making it private.

File: tests/stack/test_api.py [hunk 2]
```diff
@@ -338,3 +338,65 @@ def test_deploy_resource(self, stack_api):
...
+        test_workspace_nb_properties = TEST_WORKSPACE_NB_PROPERTIES.copy()
+        test_workspace_nb_properties.update(
+            {'source_path': os.path.join(tmpdir.strpath,
+                                         test_workspace_nb_properties['source_path'])})
+        test_workspace_dir_properties = TEST_WORKSPACE_DIR_PROPERTIES.copy()
+        test_workspace_dir_properties.update(
+            {'source_path': os.path.join(tmpdir.strpath,
+                                         test_workspace_dir_properties['source_path'])})
+
+        stack_api._download_workspace(test_workspace_dir_properties, True)
```
Comment (k4jiang) [inline tests/stack/test_api.py:490]: As a general comment, it is best not to test private methods because private methods are flexible to change. Rather, you should just test the public methods instead.

Comment (alinxie) [reply tests/stack/test_api.py:490]: oh man i have a lot of tests that test private methods... I'll note this for future developments.

File: databricks_cli/stack/api.py (hunk 1 above)
Comment (alinxie) [reply databricks_cli/stack/api.py]: i'm thinking of actually exposing this in the future and putting the code in `download` in the cli.py file.

File: databricks_cli/stack/api.py [hunk 3]
```diff
@@ -91,6 +91,29 @@ def deploy(self, config_path, **kwargs):
         click.echo("Saving stack status to {}".format(status_path))
         self._save_json(status_path, new_stack_status)
 
+    def download(self, config_path, **kwargs):
+        """
+        Downloads a stack given stack JSON configuration template at path config_path.
+
+        Loads the JSON template as well as status JSON if stack has been deployed before.
+        Changes working directory to the same directory as where the config file is, then
+        calls on download_config to do the stack downloading.
```
Comment (xuanwang14) [inline databricks_cli/stack/api.py]: The docstr here have too much implementation details, which may change in the future. The documentation here should make sense when user use "--help" to see them.

File: databricks_cli/stack/api.py [hunk 4]
```diff
@@ -91,6 +91,29 @@ def deploy(self, config_path, **kwargs):
...
+
+        :param config_path: Path to stack JSON configuration template. Must have the fields of
+        'name', the name of the stack and 'resources', a list of stack resources.
+        :return: None.
+        """
+        stack_config = self._load_json(config_path)
+        config_dir = os.path.dirname(os.path.abspath(config_path))
+        cli_dir = os.getcwd()
+        os.chdir(config_dir)  # Switch current working directory to where json config is stored
+        self.download_config(stack_config, **kwargs)
```
Comment (xuanwang14) [inline databricks_cli/stack/api.py]: Why do you have to switch dirs in order to download? Is there a way to achieve this without switching dirs?

File: databricks_cli/stack/api.py (hunk 1 above)
Comment (xuanwang14) [reply databricks_cli/stack/api.py]: The `config` part is confusing because the config file is already a local file. Maybe "_download_resources"?

File: databricks_cli/stack/api.py [hunk 5]
```diff
@@ -139,6 +162,84 @@ def deploy_config(self, stack_config, stack_status=None, **kwargs):
...
+            # Deploy resource, get resource_status
+            self._download_resource(resource_config, **kwargs)
+            click.echo('#' * 80)
+
+    def _download_resource(self, resource_config, **kwargs):
+        """
+        Downloads a resource given a resource information extracted from the stack JSON
+        configuration template.
+
+        :param resource_config: A dict of the resource with fields of RESOURCE_ID, RESOURCE_SERVICE
```
Comment (xuanwang14) [inline databricks_cli/stack/api.py]: In the documents you should use "resource_id" instead of the variable `RESOURCE_ID`. Does python do the replacements automatically?

File: databricks_cli/stack/api.py (hunk 1 above)
Comment (xuanwang14) [reply databricks_cli/stack/api.py]: Now it seems unnecessary to break `download` and `download_config`? Now both functions are very thin and it is hard to tell the differences. Maybe they can be combined into one method?

File: databricks_cli/stack/api.py [hunk 6]
```diff
@@ -139,6 +162,84 @@ def deploy_config(self, stack_config, stack_status=None, **kwargs):
...
+            )
+            overwrite = kwargs.get('overwrite_notebooks', False)
+            self._download_workspace(resource_properties, overwrite)
+        else:
+            click.echo("Resource service '{}' not supported for download. "
+                       "skipping.".format(resource_service))
+
+    def _download_workspace(self, resource_properties, overwrite):
+        """
+        Deploy workspace asset.
```
Comment (xuanwang14) [inline databricks_cli/stack/api.py]: "Deploy" -> "Download"?

File: databricks_cli/stack/api.py [hunk 7]
```diff
@@ -139,6 +162,84 @@ def deploy_config(self, stack_config, stack_status=None, **kwargs):
...
+        TODO (alinxie) Change name to overwrite to be more toward overwriting exclusive resources.
+        For a notebook workspace resource, it is recommended to provide 'language' and 'format'
+        :param resource_properties: dict of properties for the workspace asset. Must contain the
+        'source_path' and 'path' fields. The other fields will be inferred if not provided.
+        :param overwrite: Whether or not to overwrite the contents of workspace notebooks.
+        :return: (dict, dict) of (physical_id, deploy_output). physical_id is the physical ID for
```
Comment (xuanwang14) [inline databricks_cli/stack/api.py]: Why do you need return values?

File: databricks_cli/stack/api.py [hunk 8]
```diff
@@ -139,6 +162,84 @@ def deploy_config(self, stack_config, stack_status=None, **kwargs):
...
+        click.echo('Downloading {} from Databricks path {} to {}'.format(object_type,
+                                                                         workspace_path,
+                                                                         local_path))
+        if object_type == 'NOTEBOOK':
+            # Inference of notebook language and format. A tuple of (language, fmt) or Nonetype.
+            language_fmt = WorkspaceLanguage.to_language_and_format(local_path)
+            if language_fmt is None:
+                raise StackError("Workspace Notebook language and format cannot be inferred."
+                                 "Please check file extension of notebook 'source_path'.")
+            fmt = language_fmt[1]
```
Comment (xuanwang14) [inline databricks_cli/stack/api.py]: use `(fmt, _) = language_fmt`? It would be good to have comments on what are the contents of `language_fmt`

File: databricks_cli/stack/api.py [hunk 9]
```diff
@@ -139,6 +162,84 @@ def deploy_config(self, stack_config, stack_status=None, **kwargs):
...
+            local_dir = os.path.dirname(os.path.abspath(local_path))
+            if not os.path.exists(local_dir):
+                os.makedirs(local_dir)
+            self.workspace_client.export_workspace(workspace_path, local_path, fmt, overwrite)
+        elif object_type == 'DIRECTORY':
+            if not os.path.exists(local_path):
+                os.makedirs(local_path)
+            self.workspace_client.export_workspace_dir(workspace_path, local_path, overwrite)
```
Comment (xuanwang14) [inline databricks_cli/stack/api.py]: What is the behavior of this functions on folders?
(1) if overwrite, then is it a per-file overwrite. In other words, is the whole folder going to be erased first?
(2) if not overwrite, what is the behavior if (a) the folder exists but is empty (b) the folder exists and have files with different names.

File: tests/stack/test_api.py (hunk 2 above)
Comment (xuanwang14) [reply tests/stack/test_api.py:490]: I agree the tests should focus on APIs. However, I feel sometimes it is good to test helper (private) method as long as the method perform a well defined task.

File: tests/stack/test_api.py [hunk 10]
```diff
@@ -338,3 +338,65 @@ def test_deploy_resource(self, stack_api):
...
+            test_workspace_nb_properties['path']
+        assert stack_api.workspace_client.export_workspace.call_args[0][1] == \
+            test_workspace_nb_properties['source_path']
+
+        # Should raise error if object_type is not NOTEBOOK or DIRECTORY
+        test_workspace_dir_properties.update({'object_type': 'INVALID_TYPE'})
+        with pytest.raises(StackError):
+            stack_api._download_workspace(test_workspace_dir_properties, True)
+
+    def test_download_resource(self, stack_api):
```
Comment (xuanwang14) [inline tests/stack/test_api.py:511]: Replace this test with one on the public "download" method?

File: tests/stack/test_api.py [hunk 11]
```diff
@@ -338,3 +338,65 @@ def test_deploy_resource(self, stack_api):
...
+        stack_api._download_workspace(test_workspace_dir_properties, True)
+        stack_api.workspace_client.export_workspace_dir.assert_called_once()
+        assert stack_api.workspace_client.export_workspace_dir.call_args[0][0] == \
+            test_workspace_dir_properties['path']
+        assert stack_api.workspace_client.export_workspace_dir.call_args[0][1] == \
+            test_workspace_dir_properties['source_path']
+
+        stack_api._download_workspace(test_workspace_nb_properties, True)
+        stack_api.workspace_client.export_workspace.assert_called_once()
+        assert stack_api.workspace_client.export_workspace.call_args[0][0] == \
```
Comment (xuanwang14) [inline tests/stack/test_api.py]: It is hard to follow `call_args[0][0]` is there a better way to do the test? (is this the standard way to to unit tests in Databricks CLI ?)

File: tests/stack/test_api.py [hunk 12]
```diff
@@ -338,3 +338,65 @@ def test_deploy_resource(self, stack_api):
         }
         with pytest.raises(StackError):
             stack_api._deploy_resource(resource_badtype)
+
+    def test_download_workspace(self, stack_api, tmpdir):
```
Comment (xuanwang14) [inline tests/stack/test_api.py:469]: Another set of important behaviors to test are how "overwrite" is handled in different scenarios

File: databricks_cli/stack/api.py (hunk 4 above)
Comment (alinxie) [reply databricks_cli/stack/api.py]: This is the same process I'm using for deploying. I want the paths in the config file to be relative to where the config file is, so I switch to the directory of the config file. If the config file is in the current directory, then the directory isn't being switched.

File: databricks_cli/stack/api.py (hunk 5 above)
Comment (alinxie) [reply databricks_cli/stack/api.py]: Wish it did, but I put `RESOURCE_ID` since it's inline with the code. I'll switch to 'resource_id' and so on to reduce confusions. I realize this documentation should be more user-readable than developer-readable.

File: databricks_cli/stack/api.py (hunk 6 above)
Comment (alinxie) [reply databricks_cli/stack/api.py]: Yes. Actually haven't worked too much on the documentation yet, just on the code. I'll get the documentation down hopefully by monday.

File: databricks_cli/stack/api.py (hunk 7 above)
Comment (alinxie) [reply databricks_cli/stack/api.py]: will change.

File: databricks_cli/stack/api.py (hunk 8 above)
Comment (alinxie) [reply databricks_cli/stack/api.py]: okay sounds good, never thought of that syntax.

File: databricks_cli/stack/api.py (hunk 9 above)
Comment (alinxie) [reply databricks_cli/stack/api.py]: Seems like it is per-file overwrite, but I'll look into it and the cases you described.

File: tests/stack/test_api.py (hunk 11 above)
Comment (alinxie) [reply tests/stack/test_api.py]: this is how it's been done for the other unit tests. I can try finding a more understandable way to do it.

File: databricks_cli/stack/api.py (hunk 1 above)
Comment (alinxie) [reply databricks_cli/stack/api.py]: Thinking of making another PR to actually move what is in `download` and anything related to reading or saving JSON's to `cli.py` and make `download_config` the main entry endpoint of the API. Let me know what you guys think of this.

File: databricks_cli/stack/api.py (hunk 8 above)
Comment (alinxie) [reply databricks_cli/stack/api.py]: comment of what's in language_fmt at https://github.com/alinxie/databricks-cli/pull/4/files#diff-adef28b314958b556e396054d0efb708R226

File: tests/stack/test_api.py (hunk 12 above)
Comment (alinxie) [reply tests/stack/test_api.py:469]: Feel like this is something that should be handled in the `workspace` API. I planned to just check in this test that the right arguments are plugged into the `workspace` API `import_workspace` and `import_workspace_dir` functions.

File: tests/stack/test_api.py (hunk 10 above)
Comment (alinxie) [reply tests/stack/test_api.py:511]: Hm yeah that should be doable.

File: databricks_cli/stack/api.py [hunk 13]
```diff
@@ -136,6 +155,82 @@ def deploy_config(self, stack_config, stack_status=None, **kwargs):
...
+        Downloads a resource given a resource information extracted from the stack JSON
+        configuration template.
+
+        :param resource_config: A dict of the resource with fields of RESOURCE_ID, RESOURCE_SERVICE
+        and RESOURCE_PROPERTIES.
+        ex. {'id': 'example-resource', 'service': 'jobs', 'properties': {...}}
+        """
+        resource_id = resource_config.get(RESOURCE_ID)
+        resource_service = resource_config.get(RESOURCE_SERVICE)
+        resource_properties = resource_config.get(RESOURCE_PROPERTIES)
```
Comment (k4jiang) [inline databricks_cli/stack/api.py]: Something to consider for the future:
Operating on a dictionary (`resource_config`) is very difficult because it is hard to enforce a schema. As a result, there's a bunch of code in this class that validates and parses the data, which makes it difficult to identify what exactly the `_download_resource` function is doing.
 
Ideally, we would decouple how we parse the JSON elsewhere. By having this clear separation of concern, it'll make it easier to change the JSON structure and make the logic flow a lot more clear.

**Comment by k4jiang** [comment]
LGTM
